### ISSUE TITLE
init.cc:1247 NCCL WARN Mismatched collective detected

### ISSUE THREAD

**Issue by MicaelCarvalho** [NVIDIA/nccl#296]
NCCL version 2.4.8+cuda10.0

`init.cc:1247 NCCL WARN Mismatched collective detected, please check your collective calls at and around rank 39. You can use NCCL_DEBUG=INFO and NCCL_DEBUG_SUBSYS=COLL to see the collective logs`

I'm running PyTorch with DistributedDataParallel on a cluster. Adding those flags to the env does not yield any additional information to the logs -- it actually prints the exact same message suggesting I add the flags, although they are there... When running in a single GPU, everything works well. Is there any way to debug this?

This is likely one of the least informative error message I've error seen. I have zero clues on how to debug it, where it is coming from, or anything useful really.

Any suggestion is greatly appreciated.

**Comment by kwen2501** [comment]
Hi,

The collective logs may not be printed at the place where the error message is shown, they may be printed before the error.

Also, can you confirm that those two environment variables are added to all processes/nodes?

**Comment by kwen2501** [comment]
Also, this error is most likely due to providing tensors of different sizes at different ranks or enqueuing multiple collective calls in different orders at different ranks. You may also check your program in those aspects.

**Comment by MicaelCarvalho** [comment]
Confirming: The environment variables are set for all the processes/nodes.
Also checked the tensor sizes, and they were fine.
Finally, found out the problem: I was only computing losses for parts of the network I actively forward. This depends on the data being forwarded, meaning some nodes could have gradients for a few branches, and others wouldn't.

Unfortunately, the NCCL messages didn't really help... I had to spend a few hours debugging to understand what could be happening. This is solved now. Thanks @kwen2501

**Comment by sjeaugey** [comment]
As a reference (for others in the same situation), the error message is indicating that there is likely a size mismatch between different ranks, which means some ranks are calling collective operations with a different size than others, leading to a hang.

Adding the suggested environment variables would trace NCCL calls for each rank, so the user could see the difference in size between the different ranks, hopefully helping to debug the incorrect usage.

@MicaelCarvalho given you are seeing the `WARN` line (which is not the default), you might have something setting `NCCL_DEBUG=WARN` after the launch, overriding `NCCL_DEBUG=INFO`, which might be the reason why you're not seeing the call traces.

**Comment by ashwinvaswani** [comment]
@MicaelCarvalho Can you please elaborate on your fix? I am facing a similar issue, I am using a multi stream network in which I have two backbones and concatenate their outputs. I face the problem that NCCL hangs, when I use a single stream network with one backbone, it works. Can you please help!

**Comment by sjeaugey** [comment]
@ashwinvaswani This message happens when some ranks call collective operations with different sizes than others (or not at all). Setting the suggested environment variable, you can compare each rank and make sure they all call NCCL operations with the same sizes in the same order on a given NCCL communicator.

Also note if you are using ncclSend/ncclRecv you might get that message erroneously, which is fixed in [2.7.8](https://github.com/NVIDIA/nccl/commit/033d799524fb97629af5ac2f609de367472b2696).

**Comment by ashwinvaswani** [comment]
@sjeaugey Here is a link to my log file with NCCL_IB_DISABLE=1, NCCL_DEBUG=INFO, and NCCL_DEBUG_SUBSYS=ALL:
https://drive.google.com/file/d/1pncwFIzI9MJ2iy6xDyjycX92MeDsGFBF/view?usp=sharing

It is pretty verbose and I did not quite understand where ranks may be calling collective operations with different sizes. Can you please check?

**Comment by sjeaugey** [comment]
Did a bit of data processing :
```
grep "\[0\] NCCL INFO" nccl_logs.txt > log.0
grep "\[1\] NCCL INFO" nccl_logs.txt > log.1
grep "\[2\] NCCL INFO" nccl_logs.txt > log.2
grep "\[3\] NCCL INFO" nccl_logs.txt > log.3
grep "opCount" log.0 | cut -f 5-7,12-13 -d ' ' > log_0
grep "opCount" log.1 | cut -f 5-7,12-13 -d ' ' > log_1
grep "opCount" log.2 | cut -f 5-7,12-13 -d ' ' > log_2
grep "opCount" log.3 | cut -f 5-7,12-13 -d ' ' > log_3
```
Then I could see the diff the sequence :
```
vim -d log_*
+ +--640 lines: Broadcast: opCount 0 |+ +--640 lines: Broadcast: opCount 0|+ +--640 lines: Broadcast: opCount 0|+ +--640 lines: Broadcast: opCount 0
  AllReduce: opCount 280 count 512   |  AllReduce: opCount 280 count 512  |  AllReduce: opCount 280 count 512  |  AllReduce: opCount 280 count 512
  AllReduce: opCount 281 count 128   |  AllReduce: opCount 281 count 128  |  AllReduce: opCount 281 count 128  |  AllReduce: opCount 281 count 128
  AllReduce: opCount 282 count 128   |  AllReduce: opCount 282 count 128  |  AllReduce: opCount 282 count 128  |  AllReduce: opCount 282 count 128
  AllGather: opCount 283 count 1     |  AllGather: opCount 283 count 1    |  AllGather: opCount 283 count 1    |  AllGather: opCount 283 count 1
  AllGather: opCount 284 count 480   |  AllGather: opCount 284 count 480  |  AllGather: opCount 284 count 480  |  AllGather: opCount 284 count 480
  AllGather: opCount 285 count 480   |  AllGather: opCount 285 count 480  |  AllGather: opCount 285 count 480  |  AllGather: opCount 285 count 480
  AllReduce: opCount 286 count 128   |  AllReduce: opCount 286 count 480  |  AllReduce: opCount 286 count 480  |  AllReduce: opCount 286 count 480  
  AllReduce: opCount 287 count 128   |  AllReduce: opCount 287 count 480  |  AllReduce: opCount 287 count 480  |  AllReduce: opCount 287 count 480  
  AllReduce: opCount 288 count 480   |  AllReduce: opCount 288 count 128  |  AllReduce: opCount 288 count 128  |  AllReduce: opCount 288 count 128  
  AllReduce: opCount 289 count 480   |  AllReduce: opCount 289 count 128  |  AllReduce: opCount 289 count 128  |  AllReduce: opCount 289 count 128  
  AllReduce: opCount 28a count 128   |  ----------------------------------|  ----------------------------------|  ----------------------------------
  AllReduce: opCount 28b count 128   |  ----------------------------------|  ----------------------------------|  ----------------------------------
~                                    |~                                   |~                                   |~                                   
```
As you can see, rank 0 is calling two extra AllReduce operations of count 128 whereas the three other ranks don't (opCount 286 and 287 mismatch across ranks).

**Comment by ashwinvaswani** [comment]
Oh, but what can be a possible reason for this? Multi gpu training works for a single backbone MaskRCNN  network but I am facing this error when I use multiple backbones as feature extractors and pass their concatenated output to the RPN and further detector.

**Comment by sjeaugey** [comment]
That would be a question for the framework you're using. I can only tell you what I see from the NCCL API perspective.

**Comment by snash4** [comment]
@ashwinvaswani 
I work on different proble, but I faced the same issue.
After diagnosing I found that it is caused by a varying number of batches in each process. So, if one process finishes its epoch early, and others are still processing the remaining batches, the processes hang. 
ps: my each process works on a different size of data, thus have a varying number of batches.

**Comment by lawpdas** [comment]
I face the same problem when using pytorch AMP.
